include/ginkgo/core/base/batch_multi_vector.hpp

```diff
@@ -0,0 +1,800 @@
+/*******************************<GINKGO LICENSE>******************************
+Copyright (c) 2017-2023, the Ginkgo authors
+All rights reserved.
+
+Redistribution and use in source and binary forms, with or without
+modification, are permitted provided that the following conditions
+are met:
+
+1. Redistributions of source code must retain the above copyright
+notice, this list of conditions and the following disclaimer.
+
+2. Redistributions in binary form must reproduce the above copyright
+notice, this list of conditions and the following disclaimer in the
+documentation and/or other materials provided with the distribution.
+
+3. Neither the name of the copyright holder nor the names of its
+contributors may be used to endorse or promote products derived from
+this software without specific prior written permission.
+
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS
+IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
+TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A
+PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
+HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
+SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
+LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+(INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
+OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+******************************<GINKGO LICENSE>*******************************/
+
+#ifndef GKO_PUBLIC_CORE_BASE_BATCH_MULTI_VECTOR_HPP_
+#define GKO_PUBLIC_CORE_BASE_BATCH_MULTI_VECTOR_HPP_
+
+
+#include <initializer_list>
+#include <vector>
+
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/batch_dim.hpp>
+#include <ginkgo/core/base/batch_lin_op_helpers.hpp>
+#include <ginkgo/core/base/dim.hpp>
+#include <ginkgo/core/base/executor.hpp>
+#include <ginkgo/core/base/mtx_io.hpp>
+#include <ginkgo/core/base/polymorphic_object.hpp>
+#include <ginkgo/core/base/range_accessors.hpp>
+#include <ginkgo/core/base/types.hpp>
+#include <ginkgo/core/base/utils.hpp>
+#include <ginkgo/core/matrix/dense.hpp>
+
+
+namespace gko {
+
+/**
+ * BatchMultiVector is a batch matrix format which explicitly stores all values
+ * of the vector in each of the batches.
+ *
+ * The values in each of the batches are stored in row-major format (values
+ * belonging to the same row appear consecutive in the memory).
+ *
+ * @tparam ValueType  precision of matrix elements
+ *
+ * @ingroup batch_multi_vector
+ */
+template <typename ValueType = default_precision>
+class BatchMultiVector
```

Comment:
I'd also prefer BatchVector

Reply:
I prefer MultiVector for a few reasons:

1. It is semantically more accurate.
2. Will not be confused with std::vector and gko::vector etc.
3. When Dense will be split into a matrix format and a vector format later, I think that would be a MultiVector as well, and that would make this more natural.

Reply:
On 1., I think we need to balance semantic accuracy with practicability. In the most common case, people will use just a single vector. If I read the typename, the first half of it (Multi) is irrelevant to that use case, for such important types a concise name is much more useful IMO. Also in terms of implementation effort, do we (or applications) really need multivectors, or are vectors sufficient? It seems like a lot of effort went into the former, but we only did it because the rest of Ginkgo does it that way. Is that sufficient justification?
Related to naming precedents, I think we already made that decision via `distributed::Vector`. gko::vector is an internal type alias that we can freely rename, I suggested `host_vector`.

Reply:
While I agree that the common use-case is indeed a single vector, some applications definitely use multiple vectors. I think `MultiVector` captures the meaning when a single vector is used, but Vector does not for the case of MultiVectors, so if we support multiple vectors, then IMO, `Vector` is just incorrect. 

Now that I think about it and look at the `distributed::Vector` class, I think it should be renamed to `MultiVector`. Honestly, I really dislike that the C++ standard uses vector as a class name for its dynamic arrays.

On `BatchMultiVector` v/s `batch::MultiVector`, I am a bit conflicted. I like that it uniformizes the naming, but with the possibility of `using xxxxx` shortcuts, I feel like it has a lot of potential for misuse and confusion.

Reply:
I think we should be putting everything new we added here under the `batch` namespace, otherwise we are inconsistent related to `distributed`. On Vector vs. MultiVector, my opinion is not that strong, but batch::Vector or batch::MultiVector definitely reads more smoothly than BatchMultiVector, especially with the prefix being optional inside the namespace.

`using` shortcuts are fine, as long as their scope is limited. There have been enough blog posts written about why `using namespace std` is a bad habit for me to be unconcerned about this eventuality 😄